src/devices/video/vrender0.cpp

```diff
@@ -321,202 +322,130 @@ static uint16_t Alpha(QuadInfo *Quad, uint16_t Src, uint16_t Dst)
-				Offset=(Index<<6)+((ty&7)<<3)+(tx&7);\
-				if(Index==0) goto Clamped;\
-			}\
-			else\
-				Offset = ty * (Quad->TWidth) + tx;\
-\
-			if (bpp == 4)\
-			{\
-				uint8_t Texel = Quad->u.Imageb[Offset / 2];\
-				if (Offset & 1)\
-					Texel &= 0xf;\
-				else\
-					Texel = (Texel >> 4) & 0xf;\
-				Color = Quad->Pal[Texel];\
-			}\
-			else if (bpp == 8)\
-			{\
-				uint8_t Texel = Quad->u.Imageb[Offset];\
-				Color = Quad->Pal[Texel];\
-			}\
-			else if (bpp == 16)\
-			{\
-				Color = Quad->u.Imagew[Offset];\
-			}\
-			if (Color != TransColor)\
-			{\
-				if (a == 1)\
-					*pixel = Alpha(Quad, Color, *pixel);\
-				else if (a == 2)\
-					*pixel = Shade(Color, Quad->Shade);\
-				else\
-					*pixel = Color;\
-			}\
-			Clamped:\
-			++pixel;\
-			x_tx += Quad->Txdx;\
-			x_ty += Quad->Tydx;\
-		}\
-		line += Quad->Pitch;\
-		y_tx += Quad->Txdy;\
-		y_ty += Quad->Tydy;\
-	}\
-}
-TILETEMPL(16,0,0) TILETEMPL(16,0,1) TILETEMPL(16,0,2)
-TILETEMPL(16,1,0) TILETEMPL(16,1,1) TILETEMPL(16,1,2)
-
-TILETEMPL(8,0,0) TILETEMPL(8,0,1) TILETEMPL(8,0,2)
-TILETEMPL(8,1,0) TILETEMPL(8,1,1) TILETEMPL(8,1,2)
-
-TILETEMPL(4,0,0) TILETEMPL(4,0,1) TILETEMPL(4,0,2)
-TILETEMPL(4,1,0) TILETEMPL(4,1,1) TILETEMPL(4,1,2)
-
-#undef TILENAME
-#define TILENAME(bpp, t, a) \
-DrawQuad##bpp##t##a
-
-
-static void DrawQuadFill(QuadInfo *Quad)
+void vr0video_device::DrawQuadTexture(QuadInfo *Quad)
 {
-	uint32_t x, y;
-	uint16_t *line = Quad->Dest;
-	uint16_t ShadeColor = RGB32TO16(Quad->Shade);
-	for (y = 0; y < Quad->h; ++y)
+	const u32 TransColor = Quad->Trans ? RGB32TO16(Quad->TransColor) : NOTRANSCOLOR;
+	u16 *line = Quad->Dest;
+	u32 y_tx = Quad->Tx, y_ty = Quad->Ty;
+	const u32 Maskw = Quad->TWidth - 1;
+	const u32 Maskh = Quad->THeight - 1;
+	const u32 W = Quad->TWidth >> 3;
+
+	for (u32 y = 0; y < Quad->h; ++y)
 	{
-		uint16_t *pixel = line;
-		for (x = 0; x < Quad->w; ++x)
+		u16 *pixel = line;
+		u32 x_tx = y_tx;
+		u32 x_ty = y_ty;
+		for (u32 x = 0; x < Quad->w; ++x)
 		{
-			if (Quad->SrcAlpha)
-				*pixel = Alpha(Quad, Quad->Shade, *pixel);
+			u32 Offset;
+			u32 tx = x_tx >> 9;
+			u32 ty = x_ty >> 9;
+			u16 Color = ~0;
+			if (Quad->Clamp)
+			{
+				if (tx > Maskw)
+					goto Clamped;
+				if (ty > Maskh)
+					goto Clamped;
+			}
+			else
+			{
+				tx &= Maskw;
+				ty &= Maskh;
+			}
+
+			if (Quad->Tiled)
+			{
+				const u32 Index = Quad->Tile[(ty >> 3) * (W) + (tx >> 3)];
+				Offset=(Index << 6) + ((ty & 7) << 3) + (tx & 7);
+				if (Index == 0) goto Clamped;
+			}
 			else
-				*pixel = ShadeColor;
+				Offset = ty * (Quad->TWidth) + tx;
+
+			u8 Texel;
+			switch (Quad->bpp)
+			{
+			case 4:
+				Texel = Quad->u.Imageb[Offset / 2];
+				if (Offset & 1)
+					Texel &= 0xf;
+				else
+					Texel = (Texel >> 4) & 0xf;
+				Color = Quad->Pal[Texel];
+				break;
+			case 8:
+				Texel = Quad->u.Imageb[Offset];
+				Color = Quad->Pal[Texel];
+				break;
+			case 16:
+				Color = Quad->u.Imagew[Offset];
+				break;
+			default:
+				Color = ~0;
```

Comment:
Not that it changes my point about functional approach but why here now needs a default label?

Reply:
It's a case of Non-textured and/or unimplemented format.

Reply:
Chip can only render in 4bpp/8bpp/rgb565 modes, in case you wanna check against the last mode ignoring bit 0 that's a dubious debug approach to say the least.